extra/lrshift~/lrshift~.c

```diff
@@ -7,60 +7,61 @@ static t_class *lrshift_tilde_class;
 typedef struct _lrshift_tilde
 {
     t_object x_obj;
-    int x_n;
-    t_float x_f;
+    t_float  x_n;
+    t_float  x_f;
 } t_lrshift_tilde;
 
-static t_int *leftshift_perform(t_int *w)
+static t_int *lrshift_perform(t_int *w)
 {
-    t_sample *in = (t_sample *)(w[1]);
-    t_sample *out= (t_sample *)(w[2]);
-    int n = (int)(w[3]);
-    int shift = (int)(w[4]);
-    in += shift;
-    n -= shift;
-    while (n--)
-        *out++ = *in++;
-    while (shift--)
-        *out++ = 0;
-    return (w+5);
+    t_lrshift_tilde *x = (t_lrshift_tilde *) w[1];
+    t_sample *in = (t_sample *)(w[2]);
+    t_sample *out= (t_sample *)(w[3]);
+    int n = (int)(w[4]);
+    
+    //test for abs(shift) >= n and limit shift between -(n-1) and n
+    //when shift equals n, only the "*out++ = 0.0f" loop runs
+    int shift = (int)x->x_n;
+    shift = (((shift<0)?-1*shift:shift) < n)?shift:n;
+    
+    //The loops are written this way to avoid "read after write" errors
+    //when *in and *out point to the same array
+    t_sample *ptr, *in_end = in + n-1;
```

Comment:
Pd is a C89 code base, which means that *all* local variables have to be declared at the top of the block.

Reply:
Updated.